src/storage/v2/schemas.cpp

```diff
@@ -0,0 +1,79 @@
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "storage/v2/property_value.hpp"
+#include "storage/v2/schemas.hpp"
+
+namespace memgraph::storage {
+
+SchemaViolation::SchemaViolation(ValidationStatus status, LabelId label) : status{status}, label{label} {}
+
+SchemaViolation::SchemaViolation(ValidationStatus status, LabelId label, SchemaType violated_type)
+    : status{status}, label{label}, violated_type{violated_type} {}
+
+SchemaViolation::SchemaViolation(ValidationStatus status, LabelId label, SchemaType violated_type,
+                                 PropertyValue violated_property_value)
+    : status{status}, label{label}, violated_type{violated_type}, violated_property_value{violated_property_value} {}
+
+Schemas::CreationStatus Schemas::CreateSchema(const LabelId primary_label,
+                                              const std::vector<SchemaType> &schemas_types) {
+  auto res = schemas_.insert({primary_label, schemas_types}).second;
+  return res ? Schemas::CreationStatus::SUCCESS : Schemas::CreationStatus::FAIL;
+}
+
+Schemas::DeletionStatus Schemas::DeleteSchema(const LabelId primary_label) {
+  auto res = schemas_.erase(primary_label);
+  return res != 0 ? Schemas::DeletionStatus::SUCCESS : Schemas::DeletionStatus::FAIL;
+}
+
+[[nodiscard]] std::optional<SchemaViolation> Schemas::ValidateVertex(const LabelId primary_label,
+                                                                     const Vertex &vertex) {
+  // TODO Check for multiple defined primary labels
+  std::vector<std::pair<SchemaType, PropertyValue>> type_violations;
+  if (!schemas_.contains(primary_label)) {
+    return SchemaViolation(SchemaViolation::ValidationStatus::NO_SCHEMA_DEFINED_FOR_LABEL, primary_label);
+  }
+  if (!utils::Contains(vertex.labels, primary_label)) {
+    return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_HAS_NO_PRIMARY_LABEL, primary_label);
+  }
+
+  for (auto &schema_type : schemas_[primary_label]) {
+    if (!vertex.properties.HasProperty(schema_type.property_id)) {
+      return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_HAS_NO_PROPERTY, primary_label, schema_type);
+    }
+    // Property type check
+    //  TODO Can this be replaced with just property id check?
+    if (auto vertex_property = vertex.properties.GetProperty(schema_type.property_id);
+        PropertyValueTypeToSchemaType(vertex_property) != schema_type.type) {
+      return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_PROPERTY_WRONG_TYPE, primary_label, schema_type,
+                             vertex_property);
+    }
```

Comment:
Why would we keep the knowledge of which properties of schema have been matched? This way the first violation it finds it stops searcing for others and it returns a `SchemaViolation`. I am not sure I understand what you meant by deserializing only property values once?

Reply:
Why would we keep the knowledge of which schema properties have been matched? This way, the first violation it finds stops searching for others and returns a `SchemaViolation`. I am not sure I understand what you meant by deserializing only property values once?

Reply:
Currently every `GetProperty` call might deserialize some of the properties because they are serialized into a single buffer and if you want to deserialize the `n`th one, you have to kind of deserialize all of the properties before the `n`th. There I would only call `GetProperties` once and use the result of that. However I think this might be rewritten in a better when we modified the vertex creation (basically the schema can be checked before the vertex is created, so we might don't have to deal with `PropertyValue`, but only a single map of `TypedValue`/`PropertyValue`. So for now maybe it is better to leave it as is, just keep the serialization in mind.